File: app/Models/Quantity.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Quantity extends Model
{
    use SoftDeletes;

    protected $fillable = [
        'project_id',
        'work_item_id',
        'project_structure_id',
        'project_floor_id',
        'project_unit_id',
        'planned_quantity',
        'completed_quantity',
        'unit',
        'measurement_date',
        'measurement_method',
        'verified_by',
        'approved_by',
        'verified_at',
        'approved_at',
        'notes',
    ];

    protected $casts = [
        'planned_quantity' => 'decimal:2',
        'completed_quantity' => 'decimal:2',
        'measurement_date' => 'date',
        'verified_at' => 'datetime',
        'approved_at' => 'datetime',
    ];

    protected $appends = ['completion_percentage', 'remaining_quantity'];

    /**
     * İlerleme yüzdesi hesaplama
     */
    public function getCompletionPercentageAttribute(): float
    {
        if ($this->planned_quantity == 0) {
            return 0;
        }

        return round(($this->completed_quantity / $this->planned_quantity) * 100, 2);
    }

    /**
     * Kalan metraj hesaplama
     */
    public function getRemainingQuantityAttribute(): float
    {
        return max(0, $this->planned_quantity - $this->completed_quantity);
    }

    /**
     * İlişkiler
     */
    public function project(): BelongsTo
    {
        return $this->belongsTo(Project::class);
    }

    public function workItem(): BelongsTo
    {
        return $this->belongsTo(WorkItem::class);
    }

    public function projectStructure(): BelongsTo
    {
        return $this->belongsTo(ProjectStructure::class);
    }

    public function projectFloor(): BelongsTo
    {
        return $this->belongsTo(ProjectFloor::class);
    }

    public function projectUnit(): BelongsTo
    {
        return $this->belongsTo(ProjectUnit::class);
    }

    public function verifiedBy(): BelongsTo
    {
        return $this->belongsTo(User::class, 'verified_by');
    }

    public function approvedBy(): BelongsTo
    {
        return $this->belongsTo(User::class, 'approved_by');
    }

    /**
     * Hakediş ilişkisi
     */
    public function progressPayments(): HasMany
    {
        return $this->hasMany(ProgressPayment::class);
    }

    /**
     * Scopes
     */
    public function scopeForProject($query, $projectId)
    {
        return $query->where('project_id', $projectId);
    }

    public function scopeForWorkItem($query, $workItemId)
    {
        return $query->where('work_item_id', $workItemId);
    }

    public function scopeForStructure($query, $structureId)
    {
        return $query->where('project_structure_id', $structureId);
    }

    public function scopeForFloor($query, $floorId)
    {
        return $query->where('project_floor_id', $floorId);
    }

    public function scopeForUnit($query, $unitId)
    {
        return $query->where('project_unit_id', $unitId);
    }

    public function scopeVerified($query)
    {
        return $query->whereNotNull('verified_by');
    }

    public function scopeApproved($query)
    {
        return $query->whereNotNull('approved_by');
    }

    public function scopePending($query)
    {
        return $query->whereNull('verified_by');
    }

    public function scopeCompleted($query)
    {
        return $query->where('completed_quantity', '>=', \DB::raw('planned_quantity'));
    }

    public function scopeForPeriod($query, $startDate, $endDate)
    {
        return $query->whereBetween('measurement_date', [$startDate, $endDate]);
    }

    /**
     * Helper Methods
     */
    public function isCompleted(): bool
    {
        return $this->completed_quantity >= $this->planned_quantity;
    }

    public function isVerified(): bool
    {
        return $this->verified_by !== null;
    }

    public function isApproved(): bool
    {
        return $this->approved_by !== null;
    }

    /**
     * Metraj doğrulama
     */
    public function verify(int $userId): void
    {
        $this->update([
            'verified_by' => $userId,
            'verified_at' => now(),
        ]);
    }

    /**
     * Metraj onaylama
     */
    public function approve(int $userId): void
    {
        if (!$this->isVerified()) {
            throw new \Exception('Metraj önce doğrulanmalıdır.');
        }

        $this->update([
            'approved_by' => $userId,
            'approved_at' => now(),
        ]);
    }

    /**
     * Tamamlanan metrajı güncelleme
     */
    public function updateCompletedQuantity(float $completedQuantity): void
    {
        $this->update([
            'completed_quantity' => $completedQuantity,
            'measurement_date' => now(),
        ]);

        // İlgili hakediş kayıtlarını güncelle
        $this->syncProgressPayments();
    }

    /**
     * Hakediş kayıtlarını metrajla senkronize et
     */
    private function syncProgressPayments(): void
    {
        // Bu metrajın ilişkili olduğu hakediş kayıtlarını güncelle
        $this->progressPayments()->update([
            'completed_quantity' => $this->completed_quantity,
        ]);
    }
}

File: mobile_app/lib/features/quantities/domain/entities/quantity.dart
class Quantity {
  final int id;
  final int projectId;
  final String? projectName;
  final String? projectCode;
  final int workItemId;
  final String workItemName;
  final String workItemCode;
  final String workItemUnit;
  final double? workItemUnitPrice;
  final int? projectStructureId;
  final String? projectStructureName;
  final int? projectFloorId;
  final String? projectFloorName;
  final int? projectUnitId;
  final String? projectUnitCode;
  final double plannedQuantity;
  final double completedQuantity;
  final String unit;
  final DateTime measurementDate;
  final String? measurementMethod;
  final int? verifiedBy;
  final String? verifiedByName;
  final DateTime? verifiedAt;
  final int? approvedBy;
  final String? approvedByName;
  final DateTime? approvedAt;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Quantity({
    required this.id,
    required this.projectId,
    this.projectName,
    this.projectCode,
    required this.workItemId,
    required this.workItemName,
    required this.workItemCode,
    required this.workItemUnit,
    this.workItemUnitPrice,
    this.projectStructureId,
    this.projectStructureName,
    this.projectFloorId,
    this.projectFloorName,
    this.projectUnitId,
    this.projectUnitCode,
    required this.plannedQuantity,
    required this.completedQuantity,
    required this.unit,
    required this.measurementDate,
    this.measurementMethod,
    this.verifiedBy,
    this.verifiedByName,
    this.verifiedAt,
    this.approvedBy,
    this.approvedByName,
    this.approvedAt,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  // Computed properties
  double get remainingQuantity => plannedQuantity - completedQuantity;

  double get completionPercentage {
    if (plannedQuantity == 0) return 0;
    return (completedQuantity / plannedQuantity) * 100;
  }

  bool get isVerified => verifiedAt != null;
  bool get isApproved => approvedAt != null;
  bool get isPending => !isVerified;
  bool get isAwaitingApproval => isVerified && !isApproved;

  String get statusLabel {
    if (isApproved) return 'Onaylı';
    if (isVerified) return 'Doğrulandı';
    return 'Beklemede';
  }

  String get locationLabel {
    final parts = <String>[];
    if (projectStructureName != null) parts.add(projectStructureName!);
    if (projectFloorName != null) parts.add(projectFloorName!);
    if (projectUnitCode != null) parts.add(projectUnitCode!);
    return parts.isEmpty ? 'Konum belirtilmemiş' : parts.join(' - ');
  }
}
